tornado/ioloop.py

```diff
@@ -1048,7 +1050,29 @@ def start(self):
                         self._cancellations = 0
                         self._timeouts = timeouts = [timeout for timeout in timeouts
                                                      if timeout.callback is not None]
-                        heapq.heapify(timeouts)
+                        if len(timeouts) > 1:
+                            needs_heapify = True
+
+                    ntimeouts = len(timeouts)
+                    if ntimeouts > 0 and ntimeouts < 32:
```

Comment:
What happens if 'ntimeouts' is greater than 32 and 'need_heapify' remains False because the previous branch wasn't taken ? Wouldn't we end up with an unheaped array of timeouts ? If so, I'd suggest setting 'needs_heapify' to True unconditionally, and only clearing it with this new fast check.

Reply:
No, the timeouts array is already a heap. `needs_heapify` is only set (and only necessary) when the heap nature of it is destroyed somehow. In the above code, due to linear filtering of cancelled timeouts.